Versioning the retail react app

vmarta opened the pull request:
To get us ready to version the retail-react-app (or any other package in the monorepo) with its own independent version, this PR implements all of the scripts necessary to support the new release process.

Related spike PR #1167 

Remaining tasks:
- [x] Fix bugs with the publishing script ← for more details, see the todos added in the code
- [x] Undo the changes made to the generator files


# How to Test-Drive This PR
I've been testing the PR by going through the following scenarios: releasing the SDK, commerce-sdk-react, and then the retail-react-app. To thoroughly test them, please run a **local npm registry** first:

```bash
cd packages/pwa-kit-create-app
node scripts/run-local-npm.js
```

## Releasing SDK v3.0.0

1. `cd <monorepo root>`
3. `npm run bump-version -- 3.0.0-preview.0`
4. Verify that all of the packages are bumped, but independent packages like the retail-react-app would still be at their dev version but incremented (2.8.0-dev.0 → 2.8.0-dev.1)
5. `npm run bump-version -- 3.0.0`

Switching to the publishing step. Now imagine that we have pushed our changes to GitHub and CI would take care of the publishing to npm.
1. Open up the file [publish-to-npm.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/bc774ea94adb61b957ac686f5667478bdb4f2a87/scripts/publish-to-npm.js#L31) and edit the `branchName` to `release-3.0.x`.
2. Commit all the changes to have a clean working tree
3. `npm run publish-to-npm`
4. all public packages should be published.

Switching to the clean-up steps:
1. `npm run bump-version -- 3.1.0-dev` to prepare for the next minor release.
2. Commit all the changes again, so you can test the next scenario from a clean slate.

## Releasing commerce-sdk-react v1.0.0

1. `cd <monorepo root>`
2. `npm run bump-version:commerce-sdk-react -- 1.0.0-preview.0`
5. The SDK dependencies should be downgraded to 3.0.0
6. `npm run bump-version:commerce-sdk-react -- 1.0.0`

Switching to the publishing step. Now imagine that we have pushed our changes to GitHub and CI would take care of the publishing to npm.
1. Open up the file [publish-to-npm.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/bc774ea94adb61b957ac686f5667478bdb4f2a87/scripts/publish-to-npm.js#L31) and edit the `branchName` to `release-commerce-sdk-react-1.0.x`.
2. Commit all the changes to have a clean working tree
3. `npm run publish-to-npm`
4. Only the commerce-sdk-react package is published.

Switching to the clean-up steps:
1. `npm run bump-version:commerce-sdk-react -- 1.1.0-dev` to prepare for the next minor release.
2. The above step would restore the dev version of the SDK dependencies
3. Commit all the changes again, so you can test the next scenario from a clean slate.

## Releasing retail-react-app v1.0.0
These steps are pretty much the same as releasing the previous commerce-sdk-react.

1. `cd <monorepo root>`
2. `npm run bump-version:retail-react-app -- 1.0.0-preview.0`
5. The SDK dependencies should be downgraded to 3.0.0
6. `npm run bump-version:retail-react-app -- 1.0.0`

Switching to the publishing step. Now imagine that we have pushed our changes to GitHub and CI would take care of the publishing to npm.
1. Open up the file [publish-to-npm.js](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/bc774ea94adb61b957ac686f5667478bdb4f2a87/scripts/publish-to-npm.js#L31) and edit the `branchName` to `release-retail-react-app-1.0.x`.
2. Commit all the changes to have a clean working tree
4. `npm run publish-to-npm`
5. Only the retail-react-app package is published.

Switching to the clean-up steps:
1. `npm run bump-version:retail-react-app -- 1.1.0-dev` to prepare for the next minor release.
2. The above step would restore the dev version of the SDK dependencies


# Checklists

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

bfeister committed:
more fixes for out of sync package / lockfiles

yunakim714 committed:
add missing pwa-kit packages

yunakim714 committed:
Fix regex matching patterns (#1057)

* resolve package lock errors

* clean console logs and code

* more cleanup, remove duplicate items in overrideregex

* upgrade minimatch version, match all file extensions, match with no file extension

bfeister committed:
fix dependency dupes

bfeister committed:
Remove all console logs / cleanup / unnecessary code

bfeister committed:
fix some problems with final cleanup

bfeister committed:
merge

bfeister committed:
Merge branch 'feature/template-extensibility-stress-test' into feature/template-extensibility

* feature/template-extensibility-stress-test:
  fix some problems with final cleanup
  Remove all console logs / cleanup / unnecessary code
  fix dependency dupes
  add missing pwa-kit packages
  more fixes for out of sync package / lockfiles
  Add back missing `scripts` in package.json after bad merge
  Add missing template files
  fix bad upstream merge changes
  add the whole `template-retail-react-app` under `app` dir to stress test performance with high number of file overrides
  Fix `pwa-kit-dev` paths that didn't know about `overrides` in build pipeline

# Conflicts:
#	packages/_stress-test-project/build/loadable-stats.json
#	packages/commerce-sdk-react/package-lock.json
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-dev/src/configs/webpack/config.js
#	packages/pwa-kit-dev/src/configs/webpack/plugins.js
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/spike-extendend-retail-app/build/loadable-stats.json
#	packages/spike-extendend-retail-app/package-lock.json
#	packages/spike-extendend-retail-app/package.json
#	packages/template-retail-react-app/package-lock.json
#	packages/template-typescript-minimal/package-lock.json

bfeister committed:
remove stress test files post-merge

yunakim714 committed:
added pdp page override with small customization

yunakim714 committed:
add constants file to spike project and tweak overrides values

bfeister committed:
Fix bug where `<path>/dir/index.(js|jsx|etc)` was being picked up by template extensibility, but `<path>/dir/non-index-name.(js|jsx|etc)` was not

bfeister committed:
fix package.json scripts

bfeister committed:
fix falsy value defaulting causing errors in `path.join` fix broken `npm run push`

bfeister committed:
delete `build/loadable-stat.json` from git, gitignore should work from there

bfeister committed:
fix some file pathing issues in how SDK looks up default / universal / project components

bfeister committed:
Perf enhancements, don't use regex all (.*) for module replacement, reduce conditional logic execution

bfeister committed:
minor code style tweak

bfeister committed:
cleanup

yunakim714 committed:
initial work on moving code to use resolver

yunakim714 committed:
fix getoverridepath and break on ^ imports

yunakim714 committed:
add todos and comments

yunakim714 committed:
rewrite requestcontext.request for path with caret

yunakim714 committed:
perform overrideshashmap lookup for relative import

yunakim714 committed:
merge v3 and regenerate lock files with version 2

yunakim714 committed:
progress

yunakim714 committed:
fix conflicts

bfeister committed:
fix package lockfiles that were out of sync

bfeister committed:
Revert "fix package lockfiles that were out of sync"

This reverts commit 7f3c3645ef963ba4d7a7c219e66ec93ac6fe2269.

yunakim714 committed:
progress

bfeister committed:
fix out of sync lockfiles

bfeister committed:
Merge branch 'v3' into feature/template-extensibility

* v3:
  Drop npm 6 and node 14 from V3 (#1112)
  Upgrade lerna (#1111)
  add locale and currency to apiClients (#1107)
  Update test.yml (#1109)
  [V3] Invoke push action for v3 to `staging-v3` target (#1108)
  remove commerce api folder (#1105)

# Conflicts:
#	package-lock.json
#	package.json
#	packages/commerce-sdk-react/package-lock.json
#	packages/commerce-sdk-react/package.json
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/pwa-kit-create-app/package.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-dev/package.json
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/template-retail-react-app/app/mocks/mock-data.js
#	packages/template-retail-react-app/package-lock.json
#	packages/template-typescript-minimal/package-lock.json
#	packages/test-commerce-sdk-react/package-lock.json

yunakim714 committed:
merge feature branch back

yunakim714 committed:
resolve merge conflicts

bfeister committed:
Merge branch 'v3' into feature/template-extensibility

* v3:
  Sort contents of package.json files. (#1115)
  Fail fast? False! (#1116)

# Conflicts:
#	packages/pwa-kit-create-app/package.json
#	packages/pwa-kit-dev/package.json
#	packages/pwa-kit-runtime/package.json
#	packages/template-typescript-minimal/package.json

bfeister committed:
merge conflict

bfeister committed:
resolve merge conflict

bfeister committed:
update spike project's engines / regen v3 lockfiles

yunakim714 committed:
regenerate lock files

bfeister committed:
add missing node engines

bfeister committed:
add missing useToast (bad merge)

bfeister committed:
updates from underlying template-retail-react-app

bfeister committed:
drop unnecessary pages/product-detail override

bfeister committed:
Merge branch 'feature/template-extensibility' into extensibility-code-rehaul

* feature/template-extensibility:
  drop unnecessary pages/product-detail override
  updates from underlying template-retail-react-app
  add missing useToast (bad merge)
  add missing node engines
  update spike project's engines / regen v3 lockfiles
  resolve merge conflict
  merge conflict

# Conflicts:
#	packages/spike-extendend-retail-app/pwa-kit/overrides/app/components/header/index.jsx
#	packages/spike-extendend-retail-app/pwa-kit/overrides/app/pages/product-detail/index.jsx

bfeister committed:
fix dupe useToast

bfeister committed:
progress on making `charles-solution` work

bfeister committed:
initial pass at migration to `resolver.getHook('resolve').tapAsync` working, TODO next... cleanup

yunakim714 committed:
restore icons file and add extensible text to home

yunakim714 committed:
add some comments and construct hashmap in new plugin

yunakim714 committed:
working glob sync lookup??

yunakim714 committed:
resolve lookup when there is no extension

yunakim714 committed:
add component slot for testing

yunakim714 committed:
fix some lookup errors

bfeister committed:
migrate to new directory

bfeister committed:
remove unnecessary directory nesting, remove unnecessary files for "minimum file project"

bfeister committed:
add example of extending routes non-destructively, add example of quickly toggling some base constant values (e.g. categories shown on home page)

bfeister committed:
Update example route

bfeister committed:
add some demo examples of overrides

yunakim714 committed:
cleanup

bfeister committed:
commit package name change

yunakim714 committed:
rewrite requests coming from underlying template

yunakim714 committed:
cleanup

yunakim714 committed:
rename plugin file

yunakim714 committed:
resolve merge conflict

yunakim714 committed:
remove unnecessary build files

yunakim714 committed:
rename to overrides

yunakim714 committed:
fix lookup errors

bfeister committed:
Merge branch 'v3' into feature/template-extensibility

* v3:
  [Spike]Replace watch with nodemon (#1146)
  Fix Page Designer ImageWithText Link component (#1092) (#1148)
  Upgrade deprecated dependencies (#1124)
  Restart login flow if refresh_token is invalid (#1135)
  Move some util to site-utils to avoid circular imports (#1133)
  Restore old file name. (#1140)
  Update eslint configuration (#1129)
  CI: clarify the environment variables (#1127)
  Remove react-query-devtools from production build (#1121)
  Update lerna.json (#1118)
  Parallelize lighthouse ci (#1126)
  Increase test timeouts only on CI env (#1123)
  Remove unused `request` deprecated dependency. (#1125)
  Upgrade msw to latest (#1100)
  Add mergeBasket hook (#1114)
  [V3][Hooks Integration 🪝] Manually update cache for ShopperCustomer (#1113)
  dont use callback on mutateAsync (#1119)
  2-spaces not 4-spaces (#1117)

# Conflicts:
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-dev/package.json
#	packages/pwa-kit-dev/src/configs/webpack/config.js
#	packages/pwa-kit-dev/src/ssr/server/build-dev-server.js
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/template-express-minimal/package-lock.json
#	packages/template-mrt-reference-app/package-lock.json
#	packages/template-retail-react-app/package-lock.json
#	packages/template-typescript-minimal/package-lock.json
#	packages/test-commerce-sdk-react/package-lock.json

bfeister committed:
Merge branch 'feature/template-extensibility' into extensibility-code-rehaul

* feature/template-extensibility:
  [Spike]Replace watch with nodemon (#1146)
  Fix Page Designer ImageWithText Link component (#1092) (#1148)
  Upgrade deprecated dependencies (#1124)
  Restart login flow if refresh_token is invalid (#1135)
  Move some util to site-utils to avoid circular imports (#1133)
  Restore old file name. (#1140)
  Update eslint configuration (#1129)
  CI: clarify the environment variables (#1127)
  Remove react-query-devtools from production build (#1121)
  Update lerna.json (#1118)
  Parallelize lighthouse ci (#1126)
  Increase test timeouts only on CI env (#1123)
  Remove unused `request` deprecated dependency. (#1125)
  Upgrade msw to latest (#1100)
  Add mergeBasket hook (#1114)
  [V3][Hooks Integration 🪝] Manually update cache for ShopperCustomer (#1113)
  dont use callback on mutateAsync (#1119)
  2-spaces not 4-spaces (#1117)

# Conflicts:
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/pwa-kit-runtime/package.json

bfeister committed:
fix some broken upstream changes, eslint fixes

bfeister committed:
fix bad merge conflict

vmarta committed:
Bring back the deleted package.json

vmarta committed:
bring back deleted package lockfiles

vmarta committed:
Install new eslint plugin

vmarta committed:
No more relative imports in retail-react-app

vmarta committed:
Create .prettierrc.yaml

bfeister committed:
restore other deleted package lockfiles

bfeister committed:
fix some eslint issues

bfeister committed:
remove `^` imports and begin migrating logic away from previous API. TODO: get webpack to resolve `retail-react-app` base import path from within `template-retail-react-app`

bfeister committed:
short circuit plugin temporarily, get webpack `alias` working for template to resolve self-referencing `retail-react-app` imports

bfeister committed:
fix unreachable condition

bfeister committed:
wow... freaking works 🚀

bfeister committed:
cleanup, get `template-retail-react-app` working again

bfeister committed:
cleanup

vmarta committed:
Create .eslintrc.js

bfeister committed:
fix routes.jsx which no longer pulled relative template

bfeister committed:
make routes.jsx more concise

bfeister committed:
update lockfiles

bfeister committed:
add jsconfig.json for IDEs to have magic `retail-react-app` be discoverable

bfeister committed:
Merge branch 'v3' into feature/template-extensibility

* v3:
  split ssr build on local (#1155) (#1156)
  [V3] Update Page Designer (#1128)
  [V3][Hooks Integration 🪝] Replace last of the `getProps` (#1149)
  Fix Page Designer ImageWithText Link component (#1092)

# Conflicts:
#	packages/template-retail-react-app/app/components/_app-config/index.jsx

bfeister committed:
Merge branch 'feature/template-extensibility' into feature/template-extensibility-algo-refactor

* feature/template-extensibility:
  split ssr build on local (#1155) (#1156)
  [V3] Update Page Designer (#1128)
  bring back deleted package lockfiles
  fix bad merge conflict
  [V3][Hooks Integration 🪝] Replace last of the `getProps` (#1149)
  Fix Page Designer ImageWithText Link component (#1092)

# Conflicts:
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/template-express-minimal/package-lock.json
#	packages/template-mrt-reference-app/package-lock.json
#	packages/template-retail-react-app/app/components/_app-config/index.jsx
#	packages/template-retail-react-app/app/components/_app/index.jsx
#	packages/template-retail-react-app/app/page-designer/core/component/index.jsx
#	packages/template-retail-react-app/app/page-designer/core/component/index.test.js
#	packages/template-retail-react-app/app/page-designer/core/page/index.jsx
#	packages/template-retail-react-app/app/page-designer/core/page/index.test.js
#	packages/template-retail-react-app/app/page-designer/core/region/index.jsx
#	packages/template-retail-react-app/app/page-designer/core/region/index.test.js
#	packages/template-retail-react-app/app/page-designer/layouts/carousel/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid1r1c/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid2r1c/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid2r2c/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid2r3c/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid3r1c/index.jsx
#	packages/template-retail-react-app/app/page-designer/layouts/mobileGrid3r2c/index.jsx
#	packages/template-retail-react-app/app/pages/page-not-found/index.jsx
#	packages/template-retail-react-app/app/utils/test-utils.js
#	packages/template-typescript-minimal/package-lock.json
#	packages/test-commerce-sdk-react/package-lock.json

bfeister committed:
more cleanup

bfeister committed:
change package.json key from `mobify` => `ccExtensibility`

bfeister committed:
cleanup

bfeister committed:
Final cleanup, remove console.logs, make `extendable` dynamic for internal file resolution in `template-retail-react-app` (and future projects), remove some non-required files,

bfeister committed:
fix straggler non-dynamic filepath

bfeister committed:
cleanup

bfeister committed:
remove code that shouldn't be in v3+

vmarta committed:
bump-version script ignores the retail-react-app

vmarta committed:
See if this `--pre-dist-tag` works as expected

vmarta committed:
Clean up code

vmarta committed:
Add todo

bfeister committed:
sync package.json deps in extended example project

bfeister committed:
cleanup / pr feedback

vmarta committed:
Update npm scripts

vmarta committed:
Disable `npm install` step

vmarta committed:
List the packages' versions for easy review

vmarta committed:
Fix bug with versioning the root package file

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
A bit of refactoring

vmarta committed:
Some clean up

bfeister committed:
pr feedback, great catch @kevinxh

bfeister committed:
fix problem with bad copy from underlying template file

bfeister committed:
pr feedback / cleanup

bfeister committed:
cleanup

Co-authored-by: Kevin He <kevin.he@salesforce.com>

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
Update package-lock.json

vcua-mobify committed:
WIP

bfeister committed:
drop unneeded complexity

bfeister committed:
lockfiles

bfeister committed:
moving toward an array of `alias`es

vmarta committed:
Generator: try downloading retail-react-app@latest from npm

vmarta committed:
Some refactoring

bfeister committed:
Refactor template extensibility functionality to use `resolve.alias` + greatly reduce code footprint + make leading slash optional in `ccExtensibility.overridesDir`

vmarta committed:
Tweak the versioning logic

vmarta committed:
These are optional options

vmarta committed:
Use a new npm tag for testing purpose

vcua-mobify committed:
Merge branch 'feature/template-extensibility-algo-refactor' of github.com:SalesforceCommerceCloud/pwa-kit into feature/template-extensibility-algo-refactor

bfeister committed:
add brand-logo with no clash for server side viewBox

bfeister committed:
add `request-processor.js` (required for worker.js to load) and fix bad extensibility path loading in `build-dev-server.js`

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
Add some comments

vmarta committed:
Install shelljs

vmarta committed:
retail-react-app: version updates scoped to this directory only

vmarta committed:
Update versioning logic in the root

vmarta committed:
retail-react-app: simplify the versioning logic

vmarta committed:
fake sdk release v3

vmarta committed:
fake release of retail-react-app v1

bfeister committed:
lint fix

bfeister committed:
fix failing pwa-kit-react-sdk tests

bfeister committed:
cleanup

bfeister committed:
Merge branch 'v3' into feature/template-extensibility

* v3:
  fix: remove device-context and detect-device-type (#1168)
  Dependency updates (#1170)
  [Snyk] Security upgrade cosmiconfig from 7.1.0 to 8.0.0 (#1145)
  [V3] Remove `cross-fetch` from `template-retail-react-app` (#1160)
  [v3] Add suffix to ssr build files (#1158)
  Consume new version of the commerce-sdk-isomorphic to enable Phased Launches support HTTP Basic Auth (#1153)

# Conflicts:
#	packages/commerce-sdk-react/package-lock.json
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/template-retail-react-app/package-lock.json

bfeister committed:
regen package lockfiles

bfeister committed:
Merge branch 'feature/template-extensibility' into feature/template-extensibility-algo-refactor + regen lockfiles

* feature/template-extensibility:
  regen package lockfiles
  fix: remove device-context and detect-device-type (#1168)
  Dependency updates (#1170)
  [Snyk] Security upgrade cosmiconfig from 7.1.0 to 8.0.0 (#1145)
  [V3] Remove `cross-fetch` from `template-retail-react-app` (#1160)
  [v3] Add suffix to ssr build files (#1158)
  Consume new version of the commerce-sdk-isomorphic to enable Phased Launches support HTTP Basic Auth (#1153)

# Conflicts:
#	package-lock.json
#	packages/commerce-sdk-react/package-lock.json
#	packages/internal-lib-build/package-lock.json
#	packages/my-extended-retail-app/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/template-retail-react-app/app/hooks/use-einstein.test.js
#	packages/template-retail-react-app/package-lock.json
#	packages/template-typescript-minimal/package-lock.json
#	packages/test-commerce-sdk-react/package-lock.json

bfeister committed:
lint:fix

vmarta committed:
Explore publishing certain packages only

vmarta committed:
Revert "fake release of retail-react-app v1"

This reverts commit f8dcaa0bbf4be7be7ef1d663170510c44ad71205.

vmarta committed:
Revert "fake sdk release v3"

This reverts commit 3aad2dabf33ed8e257354425e15e4b6cb217a53c.

bfeister committed:
fix failing tests in CI re: `TextDecoder is not defined`

bfeister committed:
take 2 attempting to fix TextDecoder global in jest

bfeister committed:
fix jest module mapper

bfeister committed:
fix jest moduleMapper config

bfeister committed:
REALLY fix jest moduleNameMapper config for new pathing + fix node `util` global TextDecoder is not defined in jest

bfeister committed:
add brand-logo

bfeister committed:
remove previously added `AbovePDP` Template Hook component

bfeister committed:
replay #1128 from bad merge

bfeister committed:
eslint:fix for `retail-react-app` base paths

bfeister committed:
de-duplicate deps in bundle

bfeister committed:
fix path defaulting, don't pass nullish value to path.resolve

bfeister committed:
improve pathing for overridesDir in pwa-kit-dev webapack config + add missing worker/main.js

bfeister committed:
Merge branch 'v3' into feature/template-extensibility-algo-refactor

* v3:
  [v3] Add Convenience Components (#1183)
  Update commerce-sdk-react README (#1176)
  Fix missing commerce-sdk-react logout call (#1180)

# Conflicts:
#	packages/template-retail-react-app/app/components/_app/index.jsx
#	packages/template-retail-react-app/app/pages/product-list/index.jsx

bfeister committed:
fix pathing

bfeister committed:
fix pathing issue that was mysteriously UNFIXED

bfeister committed:
fix `worker.js` path

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
Now also increment the pre-release version of the templates

vmarta committed:
Now allows for publishing one package only to npm

bfeister committed:
fix app/static pathing in non-extensible builds

vmarta committed:
Make sure Lerna uses the exact versions (without carrot)

vmarta committed:
Update package.json

vmarta committed:
What I expect to see from Ben's work

vmarta committed:
Some refactoring

vmarta committed:
Scale to multiple independent packages

vmarta committed:
Make sure version updates are synced throughout

vmarta committed:
Some refactoring

vmarta committed:
Remove commented lines

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
No need to do version-pinning ourselves

Because we now call `lerna version --exact`.

vmarta committed:
Add eslint and prettier

vmarta committed:
Move script to allow reuse by future templates

vcua-mobify committed:
[Extensibility] Remove Einstein proxy and send Einstein events directly from app (#1202)

* Remove Einstein proxy and send Einstein events directly from app

* Update ssrParameters

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vcua-mobify committed:
Replace ExtendsCircularImportsPlugin with OverridesResolverPlugin (#1211)

* Replace ExtendsCircularImportsPlugin with OverridesResolverPlugin

* fix filepath problem

---------

Co-authored-by: Brian Feister <bfeister@salesforce.com>

vmarta committed:
Move `npm install` to the publishing step

vmarta committed:
Update console log

vmarta committed:
Prep for fixing a bug

vmarta committed:
Updating a package version checks deps in other packages

bfeister committed:
remove irrelevant bundlesize check to pass CI

vmarta committed:
Extract shared util functions

vmarta committed:
Consistency: add newline to EOF

vmarta committed:
Get the name of the current package

bfeister committed:
Fix `vendor.js` inclusion of `retail-react-app` files (#1198)

* fix duplicate lilbs in bundle

* fix app/static pathing for non-extensible builds

* oof, fix extensible app/static filepaths

* Resolve conflicting npm deps via `resolve.alias`

* add missing worker/main.js

* cleanup unused imports

* WIP, add list of original overridable deps

* prevent template extensibility vendor.js from treating baseline route files as deps in vendor.js

* fix irrelevant bundlesize declaration to pass CI

* drop test:max-file-size test from my-extended-retail-app

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

alexvuong committed:
replace relative with absolute path fetchTranslations (#1215)

Co-authored-by: Brian Feister <47546998+bfeister@users.noreply.github.com>

kevinxh commented on scripts/packages-with-independent-version.json:
```diff
@@ -0,0 +1 @@
+["retail-react-app"]
```
I think we should release the hook lib independently as well.

Would we be able to release independent version for the hook library by just changing this array?

vmarta replied:
Hmm I see.. currently I have an implicit assumption that only templates would have independent version. Let me make some adjustments to my work (naming, some refactoring perhaps, etc.)

vmarta replied:
@kevinxh ok, after my latest commits, this is how you would prepare a package to have its own independent version:
- Add its package name to `packages-with-independent-version.json`
- And copy and paste a few npm scripts like this:

```json
"pkg-version": "node ../../scripts/independent-pkg-version.js",
"pwa-kit-deps-version:latest": "node ../../scripts/pwa-kit-deps-version.js latest",
"pwa-kit-deps-version:sync": "node ../../scripts/pwa-kit-deps-version.js sync"
```

vcua-mobify replied:
Should we default to having commerce-sdk-react in this array?

kevinxh replied:
yes please

vmarta committed:
For now, make this package private

It's going to be deleted anyways in the near-future.

vmarta committed:
Iterate over the other packages only

vmarta committed:
Exclude these npm scripts from the smoke test

vmarta committed:
Handle scenario of an SDK package with independent version

bfeister committed:
Refactor for unit testing, fix pathing issue from old version of file parsing algoritihim

vmarta committed:
Add a 'sync' version of the script

vmarta committed:
Tweak conditions

vmarta committed:
Some refactoring

vmarta committed:
Some refactoring

vmarta committed:
Script to run a local npm registry

vmarta committed:
Add todos for bug fixes

vmarta committed:
Verify where the npm registry is

vmarta committed:
Merge branch 'feature/template-extensibility-algo-refactor' into versioning-retail-react-app

vmarta committed:
Ignore this linting error

vmarta commented on packages/my-extended-retail-app/package.json:
```diff
@@ -87,5 +87,6 @@
     "last 4 ChromeAndroid versions"
-  ]
+  ],
+  "private": "true"
```
FYI this extended-retail-app package would be deleted in another PR #1205. So in this PR, I simply make the package private.

bfeister replied:
I've deleted `my-extend-retail-app` in `feature/template-extensibility` I think we're ready to do that here as well right?

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js:
```diff
@@ -104,11 +104,15 @@ const withLocalNPMRepo = (func) => {
             // packages to it. This is safe to do – Verdaccio does not forward these
             // the public NPM repo.
             console.log('Publishing packages to the local NPM repository')
-            sh.exec('npm run lerna -- publish from-package --yes --concurrency 1 --loglevel warn', {
```
Please ignore any changes to the files in `pwa-kit-create-app`. They're just for my own testing purpose. I plan to undo the changes.

bfeister replied:
https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1209#discussion_r1203248361

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -130,7 +130,12 @@ const merge = (a, b) => deepmerge(a, b, {arrayMerge: (orignal, replacement) => r
 const runGenerator = (answers, {outputDir, verbose}) => {
     checkOutputDir(outputDir)
     // Excluding pwa-kit-create-app, these are the public pwa-kit-* packages that can be installed through NPM.
-    const npmInstallables = ['pwa-kit-react-sdk', 'pwa-kit-dev', 'pwa-kit-runtime']
```
Please ignore any changes to the files in `pwa-kit-create-app`. They're just for my own testing purpose. I plan to undo the changes.

bfeister replied:
@vmarta did you revert these now that we're aiming to finalize and merge all of the various workstreams in `feature/template-extensibility`?

vmarta commented on scripts/.eslintrc.js:
```diff
@@ -0,0 +1,10 @@
+ */
+
+module.exports = {
+    extends: [require.resolve('../packages/pwa-kit-dev/dist/configs/eslint/no-react')]
```
Adding some eslint and prettier rules for the root's scripts folder.

vmarta committed:
Update regex

vmarta committed:
Add comments

vmarta committed:
Checks for clean working tree

vmarta committed:
Handle uncommitted changes

vmarta committed:
Make sure to update the package lock files

vmarta committed:
If error, make sure to still clean up

vmarta committed:
Install from the root

vmarta committed:
Some refactoring

vmarta committed:
No longer needing this install step

Now that we've run install in earlier steps.

vmarta committed:
For safety

vmarta committed:
If not on CI, make sure to publish to local npm

In order to avoid accidental publishing to remote npm registry.

vmarta committed:
Only exit if there's an error

vmarta commented on .github/actions/publish_to_npm/action.yml:
```diff
@@ -5,16 +5,12 @@ inputs:
 runs:
   using: composite
   steps:
-  # TODO: Figure out a way to specify whether to publish to "latest" or "next" tag
```
This is resolved. Now our script would correctly assign the npm tag `latest` or `next`: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/88b26c1fcc81de25acb884e0cf77bc8063fb94b5/scripts/publish-to-npm.js#L68

vmarta commented on .github/workflows/test.yml:
```diff
@@ -281,7 +281,7 @@ jobs:
       - name: Generate ${{ matrix.template }} project
         run: |-
-          node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir generated-${{ matrix.template }}
+          node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir ${{ env.PROJECT_DIR }}
```
This is just some refactoring to make it consistent with what's happening in that file.

alexvuong commented on packages/template-retail-react-app/package.json:
```diff
@@ -16,6 +18,9 @@
     "lint:fix": "npm run lint -- --fix",
+    "pkg-version": "node ../../scripts/independent-pkg-version.js",
+    "pwa-kit-deps-version:latest": "node ../../scripts/pwa-kit-deps-version.js latest",
+    "pwa-kit-deps-version:sync": "node ../../scripts/pwa-kit-deps-version.js sync",
```
Would these scripts visible on a generated projects? I don't think it makes sense on a generated project

vmarta replied:
Hmm, I see what you mean @alexvuong . I don't mind leaving those scripts in the generated project. But we can also choose to remove them. /cc @bendvc

bfeister replied:
Let's make a decision here so we can merge @bendvc @vmarta

alexvuong replied:
I think we should remove these scripts while generating a project because they are not executable on a generated project, and a user may confused and wonder what those scripts do in the context of a generated project

kevinxh replied:
Agree with @alexvuong , it's weird to have the versioning scripts in `template-retail-react-app`, can we move these scripts to the root of monorepo?

alexvuong replied:
yea I think these scripts should stay at root

bfeister committed:
Merge branch 'develop' into feature/template-extensibility

* develop: (37 commits)
  [Phased Launch] Call Session bridge after login (#1220)
  [v3] Add multi-site suffix to auth token keys (#1208)
  Footer: fix hydration error with the locale dropdown (#1210)
  remove unused peerDependency @chakra-ui/system (#1212)
  @W-12582733: Expose env vars endpoint for E2E smoke tests (#1207)
  Upgrade to React 18 (#1166)
  Remove v3 branch name related actions (#1206)
  add test to reach test coverage threshold
  remove commerce-api folder
  Feature: Extract einstein RefArch-specific values to constant (#1200)
  Bump version number to 2.7.1 and update changelogs (#1197)
  store usid in cookies (#1193)
  make sure static files are copied on dev environment (#1196)
  [WIP] PWA Kit 2.7.1 release (#1181)
  [V2] Internal lib build typescript dev dependency (#1194)
  [V2] Re-generate lock files and fix hook lib tests (#1186)
  [v3] Add Convenience Components (#1183)
  Update commerce-sdk-react README (#1176)
  Add additional properties to ShopperLogin test types (#1185)
  Fix missing commerce-sdk-react logout call (#1180)
  ...

# Conflicts:
#	package-lock.json
#	packages/commerce-sdk-react/CHANGELOG.md
#	packages/commerce-sdk-react/package-lock.json
#	packages/internal-lib-build/package-lock.json
#	packages/pwa-kit-create-app/package-lock.json
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-react-sdk/package-lock.json
#	packages/pwa-kit-runtime/package-lock.json
#	packages/template-mrt-reference-app/package-lock.json
#	packages/template-retail-react-app/package-lock.json
#	packages/template-typescript-minimal/package-lock.json
#	packages/test-commerce-sdk-react/package-lock.json

bfeister committed:
remove demo extensible app in light of soon-to-be-merged PR from @bendvc

bfeister committed:
lockfiles from reaact18 / chakra2

alexvuong commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -145,22 +150,13 @@ const runGenerator = (answers, {outputDir, verbose}) => {
     })
 
-    extractTemplate('template-retail-react-app', outputDir)
+    // TODO: Ben (and it should grarefully handle package-not-found error)
```
Dumb question: What does fail gracefully mean in this context? the downloadPackage func does not seem to handle when an error happens?

vmarta replied:
Actually, ignore my comment there. Ben's script already exits early if there's an error. That's already good enough for me.

bfeister replied:
Let's remove the code comment if we're not doing it then

bfeister committed:
Feature/template extensibility (#1162)

* initial work on moving code to use resolver

* fix getoverridepath and break on ^ imports

* add todos and comments

* rewrite requestcontext.request for path with caret

* perform overrideshashmap lookup for relative import

* progress

* progress

* regenerate lock files

* fix dupe useToast

* progress on making `charles-solution` work

* initial pass at migration to `resolver.getHook('resolve').tapAsync` working, TODO next... cleanup

* restore icons file and add extensible text to home

* add some comments and construct hashmap in new plugin

* working glob sync lookup??

* resolve lookup when there is no extension

* add component slot for testing

* fix some lookup errors

* cleanup

* rewrite requests coming from underlying template

* cleanup

* rename plugin file

* remove unnecessary build files

* rename to overrides

* fix lookup errors

* fix some broken upstream changes, eslint fixes

* Bring back the deleted package.json

* Install new eslint plugin

* No more relative imports in retail-react-app

* Create .prettierrc.yaml

* restore other deleted package lockfiles

* fix some eslint issues

* remove `^` imports and begin migrating logic away from previous API. TODO: get webpack to resolve `retail-react-app` base import path from within `template-retail-react-app`

* short circuit plugin temporarily, get webpack `alias` working for template to resolve self-referencing `retail-react-app` imports

* fix unreachable condition

* wow... freaking works 🚀

* cleanup, get `template-retail-react-app` working again

* cleanup

* Create .eslintrc.js

* fix routes.jsx which no longer pulled relative template

* make routes.jsx more concise

* update lockfiles

* add jsconfig.json for IDEs to have magic `retail-react-app` be discoverable

* more cleanup

* change package.json key from `mobify` => `ccExtensibility`

* cleanup

* Final cleanup, remove console.logs, make `extendable` dynamic for internal file resolution in `template-retail-react-app` (and future projects), remove some non-required files,

* fix straggler non-dynamic filepath

* cleanup

* remove code that shouldn't be in v3+

* sync package.json deps in extended example project

* cleanup / pr feedback

* pr feedback, great catch @kevinxh

* fix problem with bad copy from underlying template file

* pr feedback / cleanup

* cleanup

Co-authored-by: Kevin He <kevin.he@salesforce.com>

* Update package-lock.json

* WIP

* drop unneeded complexity

* lockfiles

* moving toward an array of `alias`es

* Refactor template extensibility functionality to use `resolve.alias` + greatly reduce code footprint + make leading slash optional in `ccExtensibility.overridesDir`

* add brand-logo with no clash for server side viewBox

* add `request-processor.js` (required for worker.js to load) and fix bad extensibility path loading in `build-dev-server.js`

* lint fix

* fix failing pwa-kit-react-sdk tests

* cleanup

* lint:fix

* fix failing tests in CI re: `TextDecoder is not defined`

* take 2 attempting to fix TextDecoder global in jest

* fix jest module mapper

* fix jest moduleMapper config

* REALLY fix jest moduleNameMapper config for new pathing + fix node `util` global TextDecoder is not defined in jest

* add brand-logo

* remove previously added `AbovePDP` Template Hook component

* replay #1128 from bad merge

* eslint:fix for `retail-react-app` base paths

* de-duplicate deps in bundle

* fix path defaulting, don't pass nullish value to path.resolve

* improve pathing for overridesDir in pwa-kit-dev webapack config + add missing worker/main.js

* fix pathing

* fix pathing issue that was mysteriously UNFIXED

* fix `worker.js` path

* fix app/static pathing in non-extensible builds

* [Extensibility] Remove Einstein proxy and send Einstein events directly from app (#1202)

* Remove Einstein proxy and send Einstein events directly from app

* Update ssrParameters

* Replace ExtendsCircularImportsPlugin with OverridesResolverPlugin (#1211)

* Replace ExtendsCircularImportsPlugin with OverridesResolverPlugin

* fix filepath problem

---------

Co-authored-by: Brian Feister <bfeister@salesforce.com>

* remove irrelevant bundlesize check to pass CI

* Fix `vendor.js` inclusion of `retail-react-app` files (#1198)

* fix duplicate lilbs in bundle

* fix app/static pathing for non-extensible builds

* oof, fix extensible app/static filepaths

* Resolve conflicting npm deps via `resolve.alias`

* add missing worker/main.js

* cleanup unused imports

* WIP, add list of original overridable deps

* prevent template extensibility vendor.js from treating baseline route files as deps in vendor.js

* fix irrelevant bundlesize declaration to pass CI

* drop test:max-file-size test from my-extended-retail-app

* replace relative with absolute path fetchTranslations (#1215)

Co-authored-by: Brian Feister <47546998+bfeister@users.noreply.github.com>

* Refactor for unit testing, fix pathing issue from old version of file parsing algoritihim

* fix merge conflict resolution

* remove template extensible demo in prep for @bendvc's pr for generator work

* regen lockfiles for react 18 changes

* fix linting

---------

Co-authored-by: yunakim714 <yunakim@salesforce.com>
Co-authored-by: Vincent Marta <vmarta@salesforce.com>
Co-authored-by: Kevin He <kevin.he@salesforce.com>
Co-authored-by: vcua-mobify <vcua@salesforce.com>
Co-authored-by: vcua-mobify <47404250+vcua-mobify@users.noreply.github.com>
Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>
Co-authored-by: Alex Vuong <alex.vuong@salesforce.com>

bfeister committed:
support windows file paths

bendvc committed:
Fix issue with failed tests because of react query dev tools

bfeister committed:
Merge branch 'feature/template-extensibility' into versioning-retail-react-app

* feature/template-extensibility: (38 commits)
  support windows file paths
  Feature/template extensibility (#1162)
  lockfiles from reaact18 / chakra2
  remove demo extensible app in light of soon-to-be-merged PR from @bendvc
  [Phased Launch] Call Session bridge after login (#1220)
  [v3] Add multi-site suffix to auth token keys (#1208)
  Footer: fix hydration error with the locale dropdown (#1210)
  remove unused peerDependency @chakra-ui/system (#1212)
  @W-12582733: Expose env vars endpoint for E2E smoke tests (#1207)
  Upgrade to React 18 (#1166)
  Remove v3 branch name related actions (#1206)
  add test to reach test coverage threshold
  remove commerce-api folder
  Feature: Extract einstein RefArch-specific values to constant (#1200)
  Bump version number to 2.7.1 and update changelogs (#1197)
  store usid in cookies (#1193)
  make sure static files are copied on dev environment (#1196)
  [WIP] PWA Kit 2.7.1 release (#1181)
  [V2] Internal lib build typescript dev dependency (#1194)
  [V2] Re-generate lock files and fix hook lib tests (#1186)
  ...

# Conflicts:
#	packages/my-extended-retail-app/config/default.js
#	packages/my-extended-retail-app/config/sites.js
#	packages/my-extended-retail-app/package-lock.json
#	packages/my-extended-retail-app/package.json
#	packages/my-extended-retail-app/pwa-kit-overrides/app/assets/svg/brand-logo.svg
#	packages/my-extended-retail-app/pwa-kit-overrides/app/components/_app-config/index.jsx
#	packages/my-extended-retail-app/pwa-kit-overrides/app/components/_app/index.jsx
#	packages/my-extended-retail-app/pwa-kit-overrides/app/pages/home/index.jsx
#	packages/my-extended-retail-app/pwa-kit-overrides/app/routes.jsx
#	packages/my-extended-retail-app/pwa-kit-overrides/app/ssr.js
#	packages/my-extended-retail-app/pwa-kit-overrides/app/static/ico/favicon.ico
#	packages/my-extended-retail-app/pwa-kit-overrides/app/static/img/global/app-icon-192.png
#	packages/my-extended-retail-app/pwa-kit-overrides/app/static/img/global/app-icon-512.png
#	packages/my-extended-retail-app/pwa-kit-overrides/app/static/img/global/apple-touch-icon.png
#	packages/pwa-kit-dev/bin/pwa-kit-dev.js
#	packages/pwa-kit-dev/package-lock.json
#	packages/pwa-kit-dev/package.json
#	packages/pwa-kit-dev/src/configs/webpack/config.js
#	packages/template-retail-react-app/app/components/confirmation-modal/index.test.js
#	packages/template-retail-react-app/app/components/footer/index.jsx
#	packages/template-retail-react-app/app/components/header/index.jsx
#	packages/template-retail-react-app/app/components/list-menu/index.test.js
#	packages/template-retail-react-app/app/components/product-scroller/index.test.js
#	packages/template-retail-react-app/app/components/product-view-modal/index.test.js
#	packages/template-retail-react-app/app/components/search/index.test.js
#	packages/template-retail-react-app/app/hoc/with-registration/index.test.js
#	packages/template-retail-react-app/app/hooks/use-add-to-cart-modal.js
#	packages/template-retail-react-app/app/hooks/use-auth-modal.test.js
#	packages/template-retail-react-app/app/hooks/use-currency.test.js
#	packages/template-retail-react-app/app/hooks/use-multi-site.test.js
#	packages/template-retail-react-app/app/hooks/use-navigation.test.js
#	packages/template-retail-react-app/app/pages/account/addresses.test.js
#	packages/template-retail-react-app/app/pages/account/index.jsx
#	packages/template-retail-react-app/app/pages/account/wishlist/index.test.js
#	packages/template-retail-react-app/app/pages/account/wishlist/partials/wishlist-primary-action.test.js
#	packages/template-retail-react-app/app/pages/cart/index.test.js
#	packages/template-retail-react-app/app/pages/cart/partials/cart-secondary-button-group.test.js
#	packages/template-retail-react-app/app/pages/checkout/index.test.js
#	packages/template-retail-react-app/app/pages/checkout/partials/contact-info.jsx
#	packages/template-retail-react-app/app/pages/checkout/partials/contact-info.test.js
#	packages/template-retail-react-app/app/pages/home/index.test.js
#	packages/template-retail-react-app/app/pages/product-list/partials/empty-results.jsx
#	packages/template-retail-react-app/app/pages/product-list/partials/page-header.jsx
#	packages/template-retail-react-app/app/pages/registration/index.test.jsx
#	packages/template-retail-react-app/app/utils/site-utils.js
#	packages/template-retail-react-app/package-lock.json
#	packages/template-retail-react-app/package.json

vmarta committed:
Remove package

vmarta committed:
Make sure versions are all 3.0.0-dev

vmarta committed:
Update package lock files

vmarta committed:
Follow up to recent merge conflicts

vmarta committed:
Revert these changes to the generator

vmarta committed:
Let the output of `npm version` pass through

vmarta committed:
Revert "Let the output of `npm version` pass through"

This reverts commit 33b3058ca4633fc0648559bc72a729ba262732d9.

vmarta committed:
Make sure to `npm install` after updating dependencies

vcua-mobify commented:
Re: releasing the SDK. I see that after we bump the version in step 2, the version of template-retail-react-app goes from `3.0.0-dev` to `3.0.0-dev.0`. I'm guessing this is expected?

vcua-mobify commented:
After running bump-version to change the version to `3.0.0-preview.0`, I still see the peer dependency to `pwa-kit-dev` in pwa-kit-runtime's package-lock.json set to `3.0.0-dev`

kevinxh commented on scripts/bump-version.js:
```diff
@@ -8,43 +8,87 @@
+
+    independentPackages.forEach((pkg) => {
+        const {location, version: oldVersion} = pkg
+        // Restore and then increment to the next pre-release version
```
Is there a particular reason why we aren't leveraging lerna's [independent](https://lerna.js.org/docs/features/version-and-publish#independent-mode) mode? That way you don't need to do the "increment then restore" dance.

vmarta replied:
@kevinxh What we're aiming for is a hybrid setup that Lerna does not provide. It does things all or nothing: either all packages in the monorepo have fixed or independent version.

vmarta replied:
After discussing further, we both wonder whether we can turn everything into independent version (as implementation detail). The interface would still be the same.

vmarta replied:
@kevinxh if we turn everything into independent version, unfortunately Lerna does not allow us to programatically set the versions via the CLI.

For example:
- if `lerna version`, then it will go into interactive mode and ask us to enter in the version for each package in the monorepo. This is tedious and can be error-prone.
- if `lerna version 3.0.0`, then it will attempt to set the versions of all packages into 3.0.0.

I wish the CLI can accept multiple versions and associate them with different packages. That wish is echoed by others too in this feature request: https://github.com/lerna/lerna/issues/389

kevinxh commented on scripts/publish-to-npm.js:
```diff
@@ -0,0 +1,96 @@
+}
+
+const verifyCleanWorkingTree = () => {
+    const isWorkingTreeClean = sh.exec('git status --porcelain', {silent: true}).trim() === ''
+    if (!isWorkingTreeClean) {
+        console.error(
+            'There are some uncommitted changes. `lerna publish` expects a clean working tree.'
+        )
+        process.exit(1)
+    }
```
I find this to be really annoying when testing the generator... Is there a way to conditionally allowing publish with uncommitted changes?

vcua-mobify replied:
+1 on easing this restriction!

vmarta replied:
`lerna publish` command does not allow you to go ahead with uncommitted changes. There is no such option to do that, unfortunately.

One workaround we can try next time is to programatically commit the changes and then undo the temporary commit afterwards.

kevinxh commented:
Instead of having 2 similar commands (`bump-version` and `pkg-version`), could there maybe just one command? 

`npm run bump-version --packages=sdks --version=3.0.0-preview.0` (`sdks` as an alias for all `pwa-kit-x` packages)
`npm run bump-version --packages=template-retail-react-app,commerce-sdk-react --version=1.0.0-preview.0`

vmarta commented:
I'm looking into addressing these feedback:
- a bug with peer deps in the package _lock_ file not being in sync for some reason (thanks @vcua-mobify )
- moving the scripts and having all of them live in the monorepo root (thanks @kevinxh )

adamraya commented:
@vmarta I think we also need to consider the scenario where we release the SDK packages but we don't want to release nor make preview releases (since there are no changes) of the `retail-react-app` package.

After releasing the SDK the versions are in the state:
- `SDK@3.1.0-dev`
- `retail-react-app@3.0.0-dev.2`

But ideally, in this scenario, we are in the state:
- `SDK@3.1.0-dev`
- `retail-react-app@3.0.0-dev`

alexvuong commented on packages/template-retail-react-app/package.json:
```diff
@@ -77,7 +80,8 @@
     "react-intl": "^5.25.1",
-    "react-router-dom": "^5.3.4"
+    "react-router-dom": "^5.3.4",
+    "shelljs": "^0.8.5"
```
if we have those scripts at root, one less dependency for the generated project ; )

vmarta commented:
> @vmarta I think we also need to consider the scenario where we release the SDK packages but we don't want to release nor make preview releases (since there are no changes) of the `retail-react-app` package.
> 
> After releasing the SDK the versions are in the state:
> 
> * `SDK@3.1.0-dev`
> * `retail-react-app@3.0.0-dev.2`
> 
> But ideally, in this scenario, we are in the state:
> 
> * `SDK@3.1.0-dev`
> * `retail-react-app@3.0.0-dev`

@adamraya in that case, I initially had retail-react-app at `3.0.0-dev` (not incremented at all). But after talking with Ben last time, I changed my mind.

Now I wanted to have its version incremented too, so that there is a version of retail-react-app that got published with the latest stable or pre-release version of the SDKs.

When we release retail-react-app at 1.0.0, it will have the SDK dependency pinned at a specific version (let's say it's SDK 3.0.0). Later when the SDK is released at 3.1.0, retail-react-app will not have that yet. But at least there's a dev version published with SDK 3.1.0 as its dependency.

alexvuong commented:
An observation after checking out the scripts, I find having to run `npm run pwa-kit-deps-version:latest` after we bump version for independent package is a bit redundant. We can resolve it in either way 
1. Have the `independent-package` detect the process automatically. For example: When we run npm run pkg-version -- 1.0.0-preview.0 , it means we want to eventually publish this package soon. So if the verion does not contain dev , let’s just downgrade any dependency to latest stable version.  Otherwise, we restore the dev version
2. we can add one argument to the pkg-version script
```
npm run pkg-version -- 1.0.0-preview.0 upgrade-to-latest
npm run pkg-version -- 1.0.0 => do nothing for pwa-kit-* deps
npm run pkg-version -- 1.1.0-dev restore-dev-deps
```

alexvuong commented:
One comment about the script name `npm run pkg-version` does not give much context on what it does. There is no verb to indicate the action of the script. Maybe we can make it clearer by changing the names a little bit, say
```
"bump-sdk-version" : "node ./scripts/bump-version.js",
"bump-individual-pkg-version: "node ../../scripts/independent-pkg-version.js"
```
Either bump-individual-pkg-versions or bump-independent-pkg-version, or any word that indicates independent packages, but can take up less word is fine by me.

adamraya commented:
> When we release retail-react-app at 1.0.0, it will have the SDK dependency pinned at a specific version (let's say it's SDK 3.0.0). Later when the SDK is released at 3.1.0, retail-react-app will not have that yet. But at least there's a dev version published with SDK 3.1.0 as its dependency.

The way the generator is implemented will always use `retail-react-app@latest`, so I think currently there's no way to generate a project using those `retail-react-app@x.x.x-devX` preview versions.

See comment: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1205#discussion_r1204297075

The same constraint forces us to always release all packages, the PWA Kit packages and a new version of `retail-react-app` using in its dependencies the version of PWA Kit just released.

vmarta committed:
Prepare to consolidate all of the versioning scripts

And move them to the root of monorepo.

vmarta committed:
Move the rest of scripts into the monorepo root

vmarta committed:
commerce-sdk-react now has its own version

vmarta committed:
Use correct package name

vmarta committed:
Add todos

vmarta committed:
Reconsider where to do `npm install`

vmarta committed:
Clean up

vmarta committed:
Remove todo

vmarta committed:
Merge branch 'develop' into versioning-retail-react-app

vmarta committed:
Remove unused dependency

vmarta committed:
Some code cleanup

vmarta commented on scripts/bump-version.js:
```diff
@@ -6,45 +6,124 @@
+    independentPackages.forEach((pkg) => {
+        const {location, version: oldVersion} = pkg
+        // Restore and then increment to the next pre-release version
+        // TODO: is it possible to _not_ trigger the lifecycle scripts? See CHANGELOG.md
+        setPackageVersion(oldVersion, {cwd: location})
+        setPackageVersion('prerelease', {cwd: location})
```
There's a bug currently with the changelog file in commerce-sdk-react. Because the package has its own version, it will run into this code block, which will cause more headings to be added in the changelog file.

I'm thinking let's fix this bug in another PR, since it's not a blocker and the changelog can be manually fixed by hand.

![GitHub Desktop 2023-05-26 at 15 21 05](https://github.com/SalesforceCommerceCloud/pwa-kit/assets/847300/ba8db1f5-505f-4c7e-8d79-52586029a10e)

vmarta committed:
Merge branch 'develop' into versioning-retail-react-app

bfeister commented:
Re-iterating the known PR check issue from https://github.com/SalesforceCommerceCloud/pwa-kit/pull/1205#issuecomment-1564997708

> We are merging this without CI Windows passing. This is a known issue, but must be resolved for v3 to launch. A new GUS ticket is open here to track fixing this prior to an official / feature-complete v3 release:
> [W-13493426](https://gus.lightning.force.com/a07EE00001SLxrHYAT)

alexvuong commented on scripts/pwa-kit-deps-version.js:
```diff
@@ -0,0 +1,58 @@
+/*
```
Nit: Can we rename this file to `bump-pwa-kit-deps-version`? It is longer but give you a bit of a context what it does

alexvuong commented on scripts/independent-pkg-version.js:
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env node
```
Nit: Can we rename this file as `bump-independent-pkg-version.js`?

alexvuong replied:
Alternatively, we can put these three scripts for version bumping into a separate folder

vmarta committed:
Move the versioning scripts into their own folder

vmarta committed:
Merge branch 'develop' into versioning-retail-react-app

adamraya commented on packages/pwa-kit-create-app/scripts/run-local-npm.js:
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
```
This script is no longer in use. Should we remove it?

kevinxh commented on packages/pwa-kit-create-app/scripts/run-local-npm.js:
```diff
@@ -0,0 +1,79 @@
+
+    const cleanup = () => {
+        console.log('Shutting down local NPM registry')
+        // delete process.env['npm_config_registry']
```
🔫

vmarta committed:
Delete run-local-npm.js
